pkg/scheduler/frameworkext/framework_extender.go

```diff
@@ -96,6 +99,96 @@ func (ext *frameworkExtendedHandleImpl) NodeResourceTopologySharedInformerFactor
 	return ext.nrtSharedInformerFactory
 }
 
+type FrameworkExtender interface {
+	framework.Framework
+}
+
+type FrameworkExtenderFactory interface {
+	New(f framework.Framework) FrameworkExtender
+}
+
+type SchedulingPhaseHook interface {
+	Name() string
+}
+
+type PreFilterPhaseHook interface {
+	SchedulingPhaseHook
+	PreFilterHook(handle ExtendedHandle, state *framework.CycleState, pod *corev1.Pod) (*corev1.Pod, bool)
+}
+
+type FilterPhaseHook interface {
+	SchedulingPhaseHook
+	FilterHook(handle ExtendedHandle, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) (*corev1.Pod, *framework.NodeInfo, bool)
+}
+
+type frameworkExtenderFactoryImpl struct {
+	handle ExtendedHandle
+
+	// extend framework with SchedulingPhaseHook
+	preFilterHooks []PreFilterPhaseHook
+	filterHooks    []FilterPhaseHook
+}
+
+func NewFrameworkExtenderFactory(handle ExtendedHandle, hooks ...SchedulingPhaseHook) FrameworkExtenderFactory {
+	i := &frameworkExtenderFactoryImpl{
+		handle: handle,
+	}
+	for _, h := range hooks {
+		// a hook may register in multiple phases
+		preFilter, ok := h.(PreFilterPhaseHook)
+		if ok {
+			i.preFilterHooks = append(i.preFilterHooks, preFilter)
+		}
+		filter, ok := h.(FilterPhaseHook)
+		if ok {
+			i.filterHooks = append(i.filterHooks, filter)
+		}
+	}
+	return i
+}
+
+func (i *frameworkExtenderFactoryImpl) New(f framework.Framework) FrameworkExtender {
+	return &frameworkExtenderImpl{
+		Framework:      f,
+		handle:         i.handle,
+		preFilterHooks: i.preFilterHooks,
+		filterHooks:    i.filterHooks,
+	}
+}
+
+var _ framework.Framework = &frameworkExtenderImpl{}
+
+type frameworkExtenderImpl struct {
+	framework.Framework
+	handle ExtendedHandle
+
+	preFilterHooks []PreFilterPhaseHook
+	filterHooks    []FilterPhaseHook
+}
+
+func (ext *frameworkExtenderImpl) RunPreFilterPlugins(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod) *framework.Status {
+	for _, hook := range ext.preFilterHooks {
+		newPod, hooked := hook.PreFilterHook(ext.handle, cycleState, pod)
+		if hooked {
+			klog.V(5).InfoS("RunPreFilterPlugins hooked", "meet PreFilterPhaseHook", "hook", hook.Name(), "pod", klog.KObj(pod))
+			return ext.Framework.RunPreFilterPlugins(ctx, cycleState, newPod)
+		}
+	}
+	return ext.Framework.RunPreFilterPlugins(ctx, cycleState, pod)
+}
+
+func (ext *frameworkExtenderImpl) RunFilterPlugins(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeInfo *framework.NodeInfo) framework.PluginToStatus {
+	for _, hook := range ext.filterHooks {
```

Comment:
ext.filterHooks is a slice, but only execute first hoo, then return ext.Framework.RunFilterPlugins directly.  My question is that should ext.filterHooks be a slice or a single pointer? I prefer a slice, thus different plugin can has its own hook?

Reply:
Here extended RunFilterPlugins checks if any FilterHook confirms the hook and returns the native RunFilterPlugins.
For `Reservation`'s usage, this behavior is alright and has the same pattern with native `RunBindPlugins`.
We may standardize this framework extender later.